app/gui/src/controller/searcher/breadcrumbs.rs

```diff
@@ -26,18 +30,145 @@ impl Breadcrumbs {
+    }
+}
+
+impl From<(component::Id, Rc<Entry>)> for BreadcrumbEntry {
+    fn from((component_id, entry): (component::Id, Rc<Entry>)) -> Self {
+        let qualified_name = entry.qualified_name();
+        let displayed_name = ImString::new(&entry.name);
+        BreadcrumbEntry { displayed_name, component_id, qualified_name }
+    }
+}
+
+
+
+// ===============
+// === Builder ===
+// ===============
+
+/// A builder for the breadcrumbs list. It is used to include all parent modules when pushing the
+/// new breadcrumb to the panel.
+#[derive(Debug)]
+pub struct Builder<'a> {
+    database:   &'a model::SuggestionDatabase,
+    components: component::List,
+}
+
+impl<'a> Builder<'a> {
+    /// Constructor.
+    pub fn new(database: &'a model::SuggestionDatabase, components: component::List) -> Self {
+        Self { database, components }
+    }
+
+    /// Build a list of breadcrumbs for a specified module. The list will contain:
+    /// 1. The main module of the project.
+    /// 2. All parent modules of the [`module`].
+    /// 3. The [`module`] itself.
+    ///
+    /// Returns [`None`] if the [`module`] is not found in the database or in the components list.
+    pub fn build(self, module: &component::Id) -> Option<Vec<BreadcrumbEntry>> {
+        let mut result = Vec::new();
+        let module_name = self.components.module_qualified_name(*module)?;
+        let entry = BreadcrumbEntry::from(self.lookup(&module_name)?);
+        result.push(entry);
+        result.extend(self.collect_parents(&module_name));
+        let project_name = module_name.project_name.clone();
+        let main_module = module::QualifiedName::new_main(project_name.clone());
+        if let Some(entry) = self.lookup(&main_module) {
+            result.push(BreadcrumbEntry {
+                displayed_name: String::from(project_name.project).into(),
+                ..entry.into()
+            });
+        }
+        Some(result.reversed())
+    }
+
+    fn lookup(&self, name: &module::QualifiedName) -> Option<(component::Id, Rc<Entry>)> {
+        self.database.lookup_by_qualified_name(name.into_iter())
     }
 
-    /// Returns a currently selected breadcrumb id.
-    pub fn currently_selected(&self) -> Option<component::Id> {
-        self.currently_selected.get()
+    /// Collect all parent modules of the given module.
+    ///
+    /// Panics if the module is not found in the database.
+    fn collect_parents(&self, name: &module::QualifiedName) -> Vec<BreadcrumbEntry> {
+        let mut result = Vec::new();
+        let (_, mut current) = self.lookup(name).expect("Entry should be present in the database.");
+        while let Some(parent) = current.parent_module() {
+            if let Some((component_id, parent_entry)) = self.lookup(&parent) {
+                result.push((component_id, parent_entry.clone_ref()).into());
+                current = parent_entry;
+            } else {
+                break;
+            }
+        }
+        result
```

Comment:
We can make an iterator over all parent modules for `module::QualifiedName`. Then, this function could return `impl Iterator<Item=BreadcrumbEntry>` instead of Vec, thus avoiding allocation.

Reply:
Refactored, but returning iterator is not possible here. We need to collect parents in reverse order, which is only possible for `DoubleEndedIterator`. As `module::QualifiedName::parent_modules` also returns `impl Iterator`, and it does not implement `DoubleEndedIterator`, we need to allocate a vector here.